routing/payment_lifecycle.go

```diff
@@ -346,6 +356,33 @@ func (p *paymentLifecycle) checkTimeout() error {
 	return nil
 }
 
+// checkContext checks whether the payment context has been cancelled.
+func (p *paymentLifecycle) checkContext(ctx context.Context) error {
```

Comment:
since both this and `checkTimeout` do basically the same thing, and since they are only called from one spot and since I think we would always want to take the same failure steps no matter which one (context or timeout) causes the trigger, perhaps we can combine them instead to avoid the duplicate logic?

```
func (p *paymentLifecycle) checkTimeoutAndContext(ctx context.Context) error {
	failPayment := func(reason channeldb.FailureReason) error {
		// By marking the payment failed, depending on whether it has
		// inflight HTLCs or not, its status will now either be
		// `StatusInflight` or `StatusFailed`. In either case, no more
		// HTLCs will be attempted.
		err := p.router.cfg.Control.FailPayment(p.identifier, reason)
		if err != nil {
			return fmt.Errorf("FailPayment got %w", err)
		}

		return nil
	}

	select {
	case <-p.timeoutChan:
		log.Warnf("payment attempt not completed before timeout")
		err := failPayment(channeldb.FailureReasonTimeout)
		if err != nil {
			return err
		}

	case <-ctx.Done():
		log.Warnf("payment attempt context canceled")
		err := failPayment(channeldb.FailureReasonError)
		if err != nil {
			return err
		}

		return ctx.Err()

	case <-p.router.quit:
		return fmt.Errorf("check payment timeout got: %w",
			ErrRouterShuttingDown)

	// Fall through if we haven't hit our time limit.
	default:
	}

	return nil
}
```

Reply:
non blocking at all btw. just a suggestion - very open to being wrong here :)

Reply:
I first had the context check included in `checkTimeout` but then thought that the separation would make it clearer that we are adding an extra step in the payment status evaluation. But I agree the code dupe is not nice, so I also think its cleaner to merge the two checks.